packages/nitro-protocol/contracts/ForceMove.sol

```diff
@@ -53,13 +53,16 @@ contract ForceMove is IForceMove {
         uint8[] memory whoSignedWhat,
         Signature memory challengerSig
     ) external override {
-        bytes32 channelId = _getChannelId(fixedPart);
-
-        require(
-            (fixedPart.participants.length >= variableParts.length) && (variableParts.length > 0),
-            'challenge | You must submit at least one but no more than numParticipants states'
+        // input type validation
+        _requireValidInput(
```

Comment:
Can we consider an optional app definition of a valid state?

Reply:
couldn't that be checked inside `ForceMoveApp.ValidTransition`?

Reply:
er yes, and it _should_ be (if we make it a thing)

I think it's related to this change, but perhaps not this line?

Reply:
My point is, can't we leave that decision up to the `ForceMoveApp` developer? Thinking of it that way, it already *is* optional. I'm not sure if you are suggesting a change to *our* contracts?

Reply:
Unfortunately not, in the case that a fully-supported state is provided.

Reply:
We could recommend that a `ForceMoveApp` expose a public, pure `validState` function (that can be used internally by `validTransition` and also externally). Then, clients can optionally call or emulate `validState` before countersigning a state, particularly a state which is not a `validTransition` from any known state.

I think we aren't far off from having such functions in our examples already: 
https://github.com/statechannels/statechannels/blob/172b4529352d4af2dc9fcf8c1d7438b7de3c83f1/packages/nitro-protocol/contracts/CountingApp.sol#L16-L23

Is there are compelling reason for this check to be performed on chain? It seem pretty "edge case", to me. `checkpoint` and `conclude` are about getting a `turnNum` or `outcome` on chain. `challenge` is maybe the problematic one, since it is currently possible to get a bad state on chain which causes `validTransition` to do unexpected things during `respond`*. Is that the kind of thing you are hoping to prevent?

*of course only by all parties agreeing, which the off-chain check I am proposing would help to prevent.